composer/metrics/nlp.py

```diff
@@ -479,3 +493,143 @@ def update(self, batch: Dict[str, Any], output_logits: torch.Tensor, labels: tor
+        return client
+
+    def update(self, batch: Dict[str, Any], outputs: List[str], labels: List[str]):
+        del labels  # never used
+        client = self.get_client()
+
+        num_beams = batch['generation_kwargs']['num_beams']
+        processed_outputs = [outputs[i * num_beams:(i + 1) * num_beams] for i in range(len(batch['prompts']))]
+        for sample_outputs, sample_prompt, test_inputs, test_outputs, entry_point in zip(
+                processed_outputs, batch['prompts'], batch['test_inputs'], batch['test_outputs'],
+                batch['entry_points']):
+            self.total += torch.tensor(1.0)
+            for code_gen in sample_outputs:
+                code_gen = re.split(r'\ndef|\nclass|\n#|\nif|\nprint|\n[A-Za-z0-9]',
+                                    code_gen)[0]  # remove everything after function ends
+                final_code = sample_prompt + code_gen  # combine prompt with the code generation
+                passes_all = True
+                for test_input, test_output in zip(test_inputs, test_outputs):
+                    if client is not None:
+                        if not self.update_online_lambda(
+                                final_code, test_input, test_output, entry_point,
+                                client):  # direct evaluation of test case to the Lambda (returns whether it passes)
+                            passes_all = False
+                            break
+                    else:
+                        ret = multiprocessing.Value('b', 0)  # Store result of test case in shared memory
+                        p = multiprocessing.Process(target=self.update_offline_helper,
+                                                    args=(final_code, test_input, test_output, entry_point,
+                                                          ret))  # Evaluate test case in an independent subprocess
+                        p.start()
+                        p.join(TIMEOUT)  # wait for timeout to terminate
+                        p.terminate()
+                        if not bool(ret.value):
+                            passes_all = False
+                            break
+
+                if passes_all:
+                    self.correct += torch.tensor(1.0)
+                    break
+        if self.remote:
+            client.close()  # pyright: ignore [reportOptionalMemberAccess]
+
+    def update_offline_helper(self, code_gen: str, test_input: str, test_output: str, entry_point: str,
+                              val: multiprocessing.Value):  # type: ignore
+        # Helper function that checks a test case for accuracy
+        mod = types.ModuleType('test_module')
+        val.value = 0
+        result = None
+        expected_result = None
+        try:
+            exec(code_gen, mod.__dict__)
+
+            result = mod.__dict__[entry_point](*eval(test_input))
+            syntax_compiled = True
+            try:
+                expected_result = eval(test_output)
```

Comment:
yes, if the test output is a string, it will try to compile the string, which will fail, so that's why we have the try catch loop

e.g. 
"1" gets compiled to int:1
"[1,2,3,4] gets compiled to list: [1,2,3,4]
but "asdfjkl" needs to get compiled to str: "asdfjkl"